internal/servers/controller/listeners.go

```diff
@@ -159,56 +160,103 @@ func (c *Controller) configureForCluster(ln *base.ServerListener) (func(), error
 	return func() { go ln.GrpcServer.Serve(ln.ALPNListener) }, nil
 }
 
-func (c *Controller) stopListeners(serversOnly bool) error {
-	wg := new(sync.WaitGroup)
-
-	for i := range c.conf.Listeners {
-		ln := c.conf.Listeners[i]
-		if ln.GrpcServer != nil { // Stop Cluster gRPC Servers if any have been started.
-			wg.Add(1)
-			go func() {
-				ln.GrpcServer.GracefulStop()
-				wg.Done()
-			}()
+func (c *Controller) stopServersAndListeners() error {
+	var eg errgroup.Group
+
+	eg.Go(c.stopClusterGrpcServerAndListener)
+	eg.Go(c.stopHttpServersAndListeners)
+	eg.Go(c.stopApiGrpcServerAndListener)
+	eg.Go(c.stopAnyListeners)
+
+	return eg.Wait()
```

Comment:
@tmessi hm, that's a good point.

At this point, we could fail to stop any servers or listeners. As the process is on it's way out, _I think_ everything would terminate even if there had been a failure here when the process actually exits (although this could be OS dependent - not very familiar with how process lifecycle management works across OSes)

Regardless, now that you mention this, I think we should return _any and all_ errors for completeness, if anything. I'll change the code a bit.

Reply:
Addressed in [4580b3d](https://github.com/hashicorp/boundary/pull/1874/commits/4580b3df019b925ae947c0df7d40f5c077208a95)